### PR TITLE
Fix generic record initialization when substitution is a runtime type

### LINKED ISSUE #19118: Lists of domains and arrays don't work ("attempt to dereference nil")
### Summary of Problem

Something about our current `list` implementation doesn't seem to work for domain or array elements, where the behavior is an "attempt to dereference nil" error.

[edit: below, DavidL and Michael point out that #15929 is the root cause]

### Steps to Reproduce

**Associated Future Test(s):**
[`test/studies/adventOfCode/2021/bradc/futures/day22a-listOfDomain.chpl`](
           https://github.com/chapel-lang/chapel/blob/main/test/studies/adventOfCode/2021/bradc/futures/day22a-listOfDomain.chpl
          ) #19028 

### Configuration Information

- Output of `chpl --version`: `chpl version 1.26.0 pre-release (b94e1c4598)`

### LINKED ISSUE #20167: Nil dereference for maps of domains
Related to https://github.com/chapel-lang/chapel/issues/20164, I'm trying to cache distributed domains in Arkouda. However, I'm seeing issues when storing domains in a map. Here's a simplified reproducer using local domains.

```chpl
use Map;

type domType = {0..0}.type;
var tab: map(int, domType);

proc cachedDom(size: int) const ref {
  if !tab.contains(size) {
    tab.add(size, {0..<size});
  }
  return tab.getReference(size);
}

proc main() {
  var d = cachedDom(10);
}
```

```sh
chpl cache-dom.chpl
./cache-dom
cache-dom.chpl:8: error: attempt to dereference nil # tab.add(size, {0..<size});

CHPL_UNWIND=system chpl cache-dom.chpl --devel
./cache-dom

$CHPL_HOME/modules/internal/ChapelArray.chpl:760: error: attempt to dereference nil
Stacktrace

newRectangularDom() at $CHPL_HOME/modules/internal/ChapelArray.chpl:757
newRectangularDom() at $CHPL_HOME/modules/internal/ChapelArray.chpl:770
init() at $CHPL_HOME/modules/internal/ChapelDomain.chpl:959
cachedDom() at cache-dom.chpl:6
main() at cache-dom.chpl:13
```

Where the nil seems to be at at https://github.com/chapel-lang/chapel/blob/3ffc1fff4a2329753f5addb752beb85c1d5416c7/modules/internal/ChapelArray.chpl#L760

Note that in my real use case the domain type is either default rectangular or block depending on `CHPL_COMM`

### PR DESCRIPTION
**Pull request by dlongnecke-cray**
Fix generic record initialization when substitution is a runtime type

Attempt to recover runtime types used in the type construction of
default initialized records. This fixes some cases where a
default initialized record containing an array or domain field
would halt at runtime when said field was used.

Resolves #19118.
Resolves #20167.

The parent issue for this effort is: #15929.

Also consider: #19292, #20424

This PR does not fix all bugs related to default initialization of
aggregates that contain runtime types. The bug persists when
a type constructor containing a runtime type is obfuscated by
indirection (e.g, it is returned by a function, or uses 2+ levels of
type alias). See the added future `RuntimeTypeEscapes.chpl` for an
example.

Reviewed by @vasslitvinov. Thanks!

TESTING

- [x] `ALL` on `linux64`, `standard`

Signed-off-by: David Longnecker <dlongnecke-cray@users.noreply.github.com>

### CODE REVIEW

File: compiler/AST/astutil.cpp [hunk 1]
```diff
@@ -856,12 +856,30 @@ bool isTypeExpr(Expr* expr) {
 
     } else if (FnSymbol* fn = call->resolvedFunction()) {
       retval = fn->retTag == RET_TYPE;
+
+    // TODO: Is it safe to just check if the base expression is a type?
+    } else if (isTypeConstructorWithRuntimeTypeActual(call)) {
+      retval = true;
```
Comment (vasslitvinov) [inline compiler/AST/astutil.cpp:862]: I am curious to know the answer to that question. Can be done outside this PR, though.

File: compiler/resolution/cleanups.cpp [hunk 2]
```diff
@@ -216,6 +216,32 @@ static void removeRandomPrimitive(CallExpr* call) {
         Symbol* sym = se->symbol();
         if (isTypeSymbol(sym) || sym->hasFlag(FLAG_TYPE_VARIABLE))
           call->remove();
+
+      // Remove type construction calls that contain runtime types, now.
```
Comment (vasslitvinov) [inline compiler/resolution/cleanups.cpp:220]: Perhaps add a comment hinting at why we do not need them any more.

**Review by vasslitvinov — APPROVED** [review]
Looks good!

**Comment by dlongnecke-cray** [comment]
@bradcray In this PR I've added a user facing `USR_WARN` to the path that default-initializes a runtime type for a record field. I opted for a warning over an error because I'm not sure of how this will affect big user codes like Arkouda.

However (pending testing on Arkouda), I would like to transform this warning into an error as it causes buggy/breaking code. Does that seem appropriate?

**Comment by bradcray** [comment]
I agree that this should become an error rather than just being a warning (sooner or later, with a slight preference for sooner).  To make sure I'm understanding correctly: If user codes did hit this warning, they likely wouldn't work, or wouldn't work for the right reasons anyway, right?  From your description, it seems as though if Arkouda or CHAMPS or the like were relying on this, they'd be in trouble, right?  Or did I take too bleak a picture from it?

I definitely agree with checking on the impact of this with at least Arkouda and CHAMPS.  Since this is merged and we're running nightly test for both against master, it seems like that should just be a matter of looking for the warning in their logs?

**Comment by dlongnecke-cray** [comment]
I never thought to just look in the logs, thanks Brad. I'll do that and submit a patch making it an error tonight if it looks OK.

**Pull request merged**